sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+file `SemaSYCL.cpp`.
+
+<h3>A SYCL Kernel</h3>
+
+The SYCL constructs `single_task`, `parallel_for`, and
+`parallel_for_work_group` each take a function object or a lambda function
+ as one of their arguments. The code within the function object or
+lambda function is executed on the device.
+To enable execution of the kernel on OpenCL devices, the lambda/function object
+is converted into the format of an OpenCL kernel.
+
+<h3>SYCL Kernel Code Generation</h3>
+
+Consider a source code example that captures an int, a struct and an accessor
+by value:
+
+```C++
+constexpr size_t c_num_items = 10;
+range<1> num_items{c_num_items}; // range<1>(num_items)
+
+int main()
+{
+  int output[c_num_items];
+  queue myQueue;
+
+  int i = 55;
+  struct S {
+    int m;
+  } s = { 66 };
+  auto outBuf = buffer<int, 1>(&output[0], num_items);
+
+  myQueue.submit([&](handler &cgh) {
+    auto outAcc = outBuf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = i + s.m;
+    });
+  });
+
+  return 0;
+}
+```
+
+The input to the code generation routines is a function object that represents
+the kernel. In pseudo-code:
+
+```C++
+struct Capture {
+    Accessor outAcc;
+    int i;
+    struct S s;
+    () {
+        outAcc[index] = i + s.m;
+    }
+}
+```
+
+On the CPU a call to such a lambda function would look like this:
+```C++
+()(struct Capture* this);
+```
+       
+When offloading the kernel to a device, the lambda/function object's 
+function operator cannot be directly called with a capture object address.
+Instead, the code generated for the device is in the form of a
+�kernel caller� and a �kernel callee�.
+The callee is a clone of the SYCL kernel object.
+The caller is generated in the form of an OpenCL kernel function.
+It receives the lambda capture object in pieces, assembles the pieces
+into the original lambda capture object and then calls the callee:
+
+```C++
+spir_kernel void caller(
+    __global int* AccData, // arg1 of Accessor init function
+    range<1> AccR1,        // arg2 of Accessor init function
+    range<1> AccR2,        // arg3 of Accessor init function
+    id<1> I,               // arg4 of Accessor init function
+    int i,
+    struct S s
+)
+{
+    // Local capture object
+    struct Capture local;
+
+    // Reassemble capture object from parts
+    local.i = i;
+    local.s = s;
+    // Call accessor�s init function
+    Accessor::init(&local.outAcc, AccData, AccR1, AccR2, I);
+
+    // Call the kernel body
+    callee(&local, id<1> wi);
+}
+
+spir_func void callee(struct Capture* this, id<1> wi)
+{
+}
+```
+
+As may be observed from the example above, standard-layout lambda capture
+components are passed by value to the device as separate parameters.
+This includes scalars, pointers, and standard-layout structs.
+Certain SYCL struct types that are not standard-layout,
+such as Accessors and Samplers, are treated specially.
```

Comment:
```suggestion
Objects of certain SYCL classes - such as `accessor` and `sampler` - require translation of
some of their fields when passed from host to the kernel. For example, pointer field within
an accessor must be translated to an OpenCL or L0 memory object before it can be passed
as kernel parameter.
```